packages/react-dom/src/__tests__/ReactDOMHooks-test.js

```diff
@@ -105,6 +109,145 @@ describe('ReactDOMHooks', () => {
     expect(labelRef.current.innerHTML).toBe('abc');
   });
 
+  it('useUniqueID does not change id even if the component updates during client render', async () => {
+    const {act} = TestUtils;
+    const {useUniqueID, useState} = React;
+    let _setShowId;
+    function App() {
+      const id = useUniqueID();
+      const [showId, setShowId] = useState(false);
+      _setShowId = setShowId;
+      return (
+        <div>
+          <div aria-labelledby={id} />
+          {showId && <div id={id} />}
+        </div>
+      );
+    }
+
+    const root = ReactDOM.createRoot(container);
+    await act(async () => {
+      root.render(<App />);
+    });
+
+    expect(container.innerHTML).toMatchInlineSnapshot(
+      `"<div><div aria-labelledby=\\"c_1\\"></div></div>"`,
+    );
+
+    await act(async () => {
+      _setShowId(true);
+    });
+
+    expect(container.innerHTML).toMatchInlineSnapshot(
+      `"<div><div aria-labelledby=\\"c_1\\"></div><div id=\\"c_1\\"></div></div>"`,
+    );
+  });
+
+  it('useUniqueID: IDs match when, after hydration, a new component that uses the ID is rendered', async () => {
+    const {act} = TestUtils;
+    const {useUniqueID, useState} = React;
+    let _setShowDiv;
+    function App() {
+      const id = useUniqueID();
+      const [showDiv, setShowDiv] = useState(false);
+      _setShowDiv = setShowDiv;
+
+      return (
+        <div>
+          <div id={id}>Child One</div>
+          {showDiv && <div id={id}>Child Two</div>}
+        </div>
+      );
+    }
+
+    container.innerHTML = ReactDOMServer.renderToString(<App />);
+    const root = ReactDOM.createRoot(container, {hydrate: true});
+    await act(async () => {
+      root.render(<App />);
+    });
+    expect(container.innerHTML).toMatchInlineSnapshot(
+      `"<div data-reactroot=\\"\\"><div id=\\"s_0\\">Child One</div></div>"`,
+    );
+    await act(async () => {
+      _setShowDiv(true);
+    });
+    expect(container.innerHTML).toMatchInlineSnapshot(
+      `"<div data-reactroot=\\"\\"><div id=\\"c_0\\">Child One</div><div id=\\"c_0\\">Child Two</div></div>"`,
+    );
+  });
+
+  it('useUniqueID: IDs match when part of the DOM tree is server rendered and part is client rendered', async () => {
+    const {useUniqueID} = React;
+    let suspend = false;
+    let resolve;
+    let promise = new Promise(resolvePromise => (resolve = resolvePromise));
+
+    function Child({text}) {
+      if (suspend) {
+        throw promise;
+      } else {
+        return text;
+      }
+    }
+
+    function App() {
+      const id = useUniqueID();
+      return (
+        <div>
+          <div id={id}>Child One</div>
+          <React.Suspense fallback={'Fallback'}>
+            <div id={id}>
+              <Child text="Child Two" />
+            </div>
+          </React.Suspense>
+        </div>
+      );
+    }
+
+    container.innerHTML =
+      '<div data-reactroot=""><div id="s_0">Child One</div><!--$!--><div>Fallback</div><!--/$--></div>';
```

Comment:
You should also assert that React didn't replace the original DOM node. I would grab a reference to the divs here, then after you hydrate, make sure they match the hydrated ones (you can use a ref for that part, or grab them out of the container again). Then do the same after the update.